ENH: GLM Tweedie Family

thequackdaddy opened the pull request:
Hello,

This will (eventually) resolve #2858 and part of #2804 . 

This adds Tweedie to the family of distributions. 

A few shortcomings...

* I don't have Stata so none of my checks are from Stata. (Also, I'm not sure where the ansocombe residuals came from. Perhaps this is a stata thing too?)
* As the deviance function doesn't like it when endog=0 or the shape parameter is 1 or 2, I've set endog to have a min of 0.1 and used WolframAlpha to derive the deviance function for the special case shape parameters (1 and 2). 
* All the residuals seem to match except working residuals when var_power = 1 and link_power = 0 (i.e., the log link)
* I don't really know how to write the loglikihood function for this. Its in the Glenn Myers PowerPoint and some papers, but haven't cracked it yet. 
* The documentation is (of course) incomplete. As I get a get a better handle for how this will work, I'll start writing it. 

So let's get rockin.

---
Updated 3/28/2016

thequackdaddy commented:
Getting closer... not there yet. Working residual isn't working for 1 of my test cases, don't have anscombe. 

I skimmed through the Tweedie LLF R function and was quite confused exactly how that function worked. Gonna take a look at it again. I'd like to have loglikelihoods, but deviance is better than nuthin'.

josef-pkt commented:
@thequackdaddy    We are not allowed to translate GPL licensed code!  GPL is incompatible for use in a BSD licensed package.

thequackdaddy commented:
@josef-pkt Yes, I realized that after the fact. That copied code was removed. Instead, I used the deviance function and calculated the limit of the function as p-> 1 and p-> 2. The answer appears to be the same. Also, I think its slightly shorter, so that's nice.

Apologies for the confusion.

josef-pkt commented:
This looks already very good based on the passing tests.

Note, if loglike is too difficult to implement right now, then we should raise an exception in the `_fit_gradient`. 
Most tests on fit can just be done with the deviance which looks available.

I doubt I have time to read it carefully today. I only briefly browsed the code yesterday.

thequackdaddy commented:
@josef-pkt Great. Any thoughts on anscombe residuals? I'll confess I don't use them much so I have little idea how to interpret them. I generally use pearson/deviance. 

Let me know if you can figure out the formula. Also, if you could get them from stata (or R if that's possible), I'd appreciate it.

josef-pkt commented:
anscombe residuals are family specific formulas (where I never looked at the details)
I was reading SAS or SPSS documentation in the past, but don't remember if they have anscombe formulas.
Given that deviance worked well, I assume anscombe should also have a reasonable expression.
But it's not a strictly necessary feature. (Main usage, AFAIK, is for outlier diagnostic) Compared to the other residuals (Pearson, deviance) it also corrects for skewness.

josef-pkt commented:
no luck on anscombe residuals in a bit of search
SPSS genlin has a good technical documentation, but I didn't see anscombe residuals (they have all the other ones)
SAS stata also doesn't seem to have anscombe, SAS insight has anscombe but no Tweedie, AFAICS
Stata doesn't have Tweedie.

thequackdaddy commented:
@josef-pkt This may belong is a separate issue, but I couldn't find anything. 

Question about "working residuals", as far as I can tell, this appears to be an "r-specific" thing. I can't find any literature talking about this concept. The best I can find is [here](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/glm.html) which suggests they are simply the residuals from the last IRLS fit. 

R appears to define them as `residuals <-  (y - mu)/mu.eta(eta)`. `mu.eta` is the derivative of mu with repect to to eta. For log, mu = exp(eta) -> dmu / deta = exp(eta). 

However, we define working residuals as and basically the [inverse](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/generalized_linear_model.py#L1167). Specifically, I think we do `(endog - mu) / sm.family.link.deriv(mu)`. Here, our denominator is the derivative of eta with respect to mu, which means eta = ln(mu) -> deta / dmu = 1 / mu. 

The point of all this is that I remain unable to match R's working residuals, although I think that's just more of a result of the way R does the residuals vs. the way we do them. With that said, the working residuals appears to have been used many times before in testing. I suppose that the working residuals are identical when you have the identity link function, but under any any function, my guess is they'd be different.  

Alternatively, it could just be there's no clear definition of the concept. With that said, intuitively, `(y -mu)/mu.eta` makes intuitive sense to me. (the response residual divided by the slope of mu at eta is basically the sensitivity of the residual with respect to eta. On the other hand, `(endog - mu) / self.family link.deriv(mu)` is a mystery to me. That's trying to find a relationship between the response residual and eta, but it doesn't seem to be accounting for a link function. 

So anyways, I'm confused by this.

josef-pkt commented:
I never looked at working residuals.
When in doubt I usually check the unit tests to see what they compare against, and comments Skipper added when he wrote the tests.
http://www.stata.com/manuals14/rglmpostestimation.pdf page 6
The glm_test_resids.py file says all residuals were obtained with Stata.

However, I would interpret the Stata manual as `(endog - mu) * self.family link.deriv(mu)`

thequackdaddy commented:
@josef-pkt Ok after looking at this more...

1.) I remain fairly confident that the way we are doing it now is incorrect. (Perhaps stata had a bug in their old version of their code and its now more apparent?)
2.) Working residuals has another problem i think. We are applying the `_n_trials` and `_freq_weight` twice. First, we apply them when calculating [resid_response](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/generalized_linear_model.py#L1157) and then again when calculating [resid_working](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/generalized_linear_model.py#L1167). I'm assuming our tests just didn't notice that. 

With #2 in mind, I seem to recall that you began leaning (towards the end of my commits for `freq_weights` to not have weights applied to the residuals. I didn't really care then, but now see that's probably a bad idea. If its alright with you, I'd like to remove weights from residuals before I proceed further with Tweedie. Further note: should I remove both `freq_weights` and `n_trials` or only `freq_weights`? 

Our algorithm does not match Stata's documentation nor the R documentation. Perhaps Stata had a bug which is now fixed? I can't test as I don't have it.

josef-pkt commented:
I have Stata, but need to clean out previous work before I can run it on an example for this.
The Stata manual wasn't publicly accessible when Skipper wrote this, and didn't have Stata or the documentation back then.

Yes, I think it's better to leave the weights out of the residuals (remove them again).
For binomial n-trials it would be best to have the status before the GLM weights, then it wouldn't be a backwards incompatible change. (My guess is that counts, i.e. multiplied by n trials would make more sense.)

josef-pkt commented:
I get different working residuals than those in the test suite after running the cpunish poisson glm example in Stata. Also, those working residuals in the test suite are much larger than any of the other residual. 

@jseabold  Do you remember why the working residuals might be wrong?

I don't find any changelog or bug reports for Stata, so I cannot check what it looked like before Stata 12

josef-pkt commented:
I think I have anscombe residuals.
An easy explanation is in "Modern Actuarial Risk Theory - Using R" page 305 equ (11.19) and following.
Generalizing this to variance function in the form `V=mu**p`, it should be

```
c = (3 - p) / 3
resid_anscombe = c * (y**c - mu**c) / mu**(p / 6)
```

which should specialize to Poisson with p=1 and gamma for p=2, gaussian for p=1
inverse gaussian has p=3 and they have log(y) instead of y**c and mu**c, i.e.
`resid_anscombe = (log(y) - log(c)) / mu**2`

(I didn't try to check the limit as p -> 3 in the derivation.)

thequackdaddy commented:
@josef-pkt 

Going back to Anscombe... 

do you mean this?

```python
    def resid_anscombe(self, endog, mu):
        """
        The Anscombe residuals for the Tweedie family

        Parameters
        ----------
        endog : array-like
            Endogenous response variable
        mu : array-like
            Fitted mean response variable

        Returns
        -------
        resid_anscombe : array
            The Anscombe residuals as defined below.

        Notes
        -----
        Not yet...
        """
        c = (3 - self.link_power) / 3
        if self.link_power == 3:
            return (np.log(endog) - np.log(c)) / mu ** 2
        else:
            return c * (endog ** c - mu ** c) / mu ** (self.link_power / 6)
```

I don't understand `resid_anscombe = (log(y) - log(c)) / mu**2` when p == 3 (inverse gaussian). c in that case is 0, so np.log(c) = np.nan.

thequackdaddy commented:
Ignore this last comment... In mixing var and link power...

josef-pkt commented:
No, I think this has to be the variance power.
the link doesn't show up it's hidden behind the mu, and this is supposed to work for any valid links.

josef-pkt commented:
I think I have a type there, if c=3, the expression should have mu not c
`resid_anscombe = (log(y) - log(mu)) / mu**2`

i.e. residuals of log transformed endog and fittedvalues.
I can check the reference again.
It's also supposed to be hardcoded in the inverse gaussian family.

josef-pkt commented:
yes, InverseGaussian family class has anscombe
`np.log(endog / mu) / np.sqrt(mu)`

BTW: np.log(0) is -inf

There are problems if endog=0, but inverse gaussian is continuous with support endog > 0, AFAIK
which I think is true for all Tweedie with power p >= 2

thequackdaddy commented:
@josef-pkt 

Back to anscombe...

Is there another typo in this? For inverse gaussian, I think we want sqrt(mu) in the denominator. For the other cases, I think we want to multiply by (1 / c), not by c. So something like this... 

```python
        if self.var_power == 3:
            return (np.log(endog) - np.log(mu)) / np.sqrt(mu)
        else:
            c = (3 - self.var_power) / 3
            return (1 / c) * (endog ** c - mu ** c) / mu ** (self.link_power / 6)
```

This should become... 
p == 0 (Gaussian) `endog - mu`
p == 1 (Poisson) `(3 / 2) * (endog ** (2 / 3) - mu ** (2 / 3)) / mu ** (1 / 6)`
p == 2 (Gamma) `3 * (endog ** (1 / 3) - mu ** (1 / 3)) / mu ** (1 / 3)`
p == 3 (Inverse Gaussian) `np.log(endog / mu) / np.sqrt(mu)`

josef-pkt commented:
What you have now looks correct.
The cases are the same as in the book.
(I don't know where my notes with the derivation are.)
That's why I like unit tests, I'm too prone to small calculation mistakes and typos. (Fortunately small calculation mistake often produce widely incorrect numbers that are easy to detect, e.g. reversing a sign.)

thequackdaddy commented:
I'm totally with you on unit tests... I've found them extremely helpful.  That's actually how I found this. 

We should be able to validate that Tweedie's residuals match for p == 0, 1, 2, and 3. I guess we'll have to assume that the other values would match as well cause I don't have a source for any other p. (Specifically, I can't get anscombe residuals from R. I did find This package called "surveillance" which advertises anscombe residuals, but they appear to be poisson-specific and then they are standardized (for reasons I don't understand). 

Thanks.

josef-pkt commented:
I think if anscombe residuals agree with the special cases, then there is no mistake left in the formula, and it's sufficient for testing (only p=3 is a special case, all others use the same formula).

thequackdaddy commented:
Well I'm struggling again... 

Probably need to take a break for today

I made an attempt at writing a function that replicated formula 3.9 in the GLM code. But I couldn't for the life of me make it stable, nor match any result I can get in R.... 

Take this code (stripped down from an attempted test I wrote)...

```python
import numpy as np
import statsmodels.api as sm
p_orig = 1.
p_new = 1.5

data = sm.datasets.cpunish.load_pandas()
count = 0
while np.abs(p_new - p_orig) > 1e-6:
    count += 1
    p_orig = p_new
    family_link = sm.families.Tweedie(var_power=p_new, link_power=0.)
    model = sm.GLM(endog=data.endog, exog=data.exog[['INCOME', 'SOUTH']],
                   family=family_link)
    res = model.fit()
    p_new = model.tweedie_estimate_power(res.mu, low=1.01, high=10)
    
x = res.cov_params().copy().values
res = model.fit(cov_type='HC0')
y = res.cov_params().values
```
p eventually stabilizes after over 40 fits to aroudn 2.40. 40 Fits just seems way to many. I found that annoying, but acceptable. 

Additionally, (and I might be confused exactly the test you wanted), I tried to validate that cov_params matched with an "optimized" var_power and with `cov_type='HC0'`. I was unsuccessful there as well. 

The real problem is when i tried to load real data that I have into the data. Again, this is insurance data so most policies (rows) have endog of 0, while the rest are some positive claim amount in the thousands or so. In my experience, p should be between 1 and 2, and probably closer to 1.6-1.7. I was again taking lots and lots of fits to get to converge, and p was over 2.7 and rising when I gave up. 

So this is a disappointment.

josef-pkt commented:
Do you get reasonable results with R on your actual data?

If you provide the same tweedie power parameter as is estimated or fixed in R, do you get the same results for params, approximately bse and deviance as R?

Can you write a function to transform parameter between tweedie and compound poisson representation? 
That would be useful for the interpretation, and for writing simulated cases to check what's going on and how robust it is.

The optimization problem for p might be messier than I thought. I think the solution has to be restricted in this case to bounded inside (1, 2). power >= 2 doesn't have zeros anymore in the tweedie distribution. It could still make sense in terms of overdispersion, but that isn't the prime focus.

I'll try to look at it, but don't have much time left today.

thequackdaddy commented:
Thanks for your comments Josef. Again, I really appreciate all your help. 

As for the data, its quite large... I have 1.2 million records--this (sadly) represents only about 10% of the data I really want to use. I can fit a GLM on this now with statsmodels in around 45 seconds (around 15 seconds to convert a dask dataframe to a design matrix and 30 seconds of statsmodels fitting time... again, statsmodels never meets a convergence tolerance, so I limit it to 10 iterations. Looking at `fit_history['deviance']` suggests the model is stable regardless. In any event, statsmodels fit times are amazing.

As for R, I've kicked it off, but numpy's matrix multiplication is much faster than R (both off the shelf R and Microsoft's MKL optimized R, as far as I can tell). I've exported my exog and endog in a CSV file and wrote the following code...

```r
setwd('c:/workspace')
data <- read.csv('output.csv', sep=",")

library(statmod)
library(tweedie)

tweedie.profile(Response ~ X1 + X2 + X3 + X4 + X5, p.vec=seq(1, 2, length=21), link.power=0, data=data, do.plot=TRUE)
```
`X1 + X2 + X3 + X4` are just transformed variables from my python code... there's actually about 25 of them, but R struggles...

Its running in Microsoft's R Open thing which is supposed to have MKL optimizations... My hope is it will finish sometime this decade. We'll see. I'll let you know... It might not finish by the time I leave work today.

josef-pkt commented:
Sounds good, thanks for the feedback. We don't get much feedback on performance with a bit larger sample sizes.

I briefly went over the unit tests that you have, and this looks pretty good.
Except for the missing loglikelihood function tweedie is in about the same stage as negativebinomial with fixed extra parameter.
The link interface needs to be adjusted to follow the same pattern as the other families, otherwise I didn't see anything in a quick look. 
This part looks about ready for final review and merge. Estimating the dispersion parameters still needs to be resolved as you mentioned earlier. I might "play" a bit with it when the kids are in bed.

thequackdaddy commented:
For loglikelihood I was intentionally holding off... 

R doesn't show logliklihood for all the quasi types. I seem to recall the reason being that R is written by purists and they want to emphasize that loglikelihood != quasi loglikelihood (which I presume wouldn't match extended quasi loglikelihood). 

So I'm not sure what I'd want there just yet... 

I was leaning towards the extended quasi loglikelihood for simplicity of programming. But I don't have a way of unit testing that. (R doesn't seem to use extended q ll).

Considering the mess that the whole infinite sum problem poses, I'm probably against having loglikelihood be anything that complex. I'd want the ability to do the infinite sum if that turns out to be the best way to estimate var_power... But use it only for that explicit purpose.

josef-pkt commented:
R has the full (infinite sum approximation) distribution and loglikelihood available, but I haven't checked what's in the GLM function of the packages. If that is maximized also for p, then it would be full MLE for tweedie, not quasi- or quasi-extended GLM.

If/When we get the "pearson function" estimator for the dispersion parameters to work, then there is essentially no need for the loglikelihood, and we should be much faster than anything that uses a full likelihood (truncated infinite sum).
In contrast, NegativeBinomial has a pretty cheap likelihood function, and it can be also directly used to estimate the dispersion parameter. For example (I looked for it earlier today for NegativeBinomial), Kerby has an iteration to maximize loglike with respect to the dispersion parameter at the end of this notebook http://nbviewer.jupyter.org/urls/umich.box.com/shared/static/0lgt5635uo6rhakjj9d5.ipynb
just evaluating on some grid.

thequackdaddy commented:
Thanks for Kerby's notebook. Very cool. 

As for logliklihood, I was merely referring to what happens in the "summary" screen. I believe R's summary function won't calculate any loglikihood for any quasi model... it just skips it. To calculate it, I thought you had to run to run dtweedie (or logLik) function, depending on the family. 

It would be nice to have an integrated approach, as to all this fragmented stuff R has.

josef-pkt commented:
I briefly tried out the gist example again, changing it to a linear mu. In that example it seems to work well.

We need some diagnostics plots. They help at least to understand how the things look like.
I tried to estimate the log OLS version for the power parameter, but it doesn't look close.
using GLM Gamma looks close to the brentq solution in the example

```
>>> res_glm = GLM((endog - mu)**2, np.column_stack((np.ones(len(endog)), np.log(mu))), family=families.Gamma(link=links.log)).fit()
>>> res_glm.params
array([ 7.23160724,  1.29238253])
>>> np.exp(res_glm.params[0])
1382.4426404781152
>>> print(power, scale)   # from Pearson estimating functions
(1.2923825491346437, 1382.4425078036968)
```

which would confirm the comment in the paper that Pearson is equivalent to Gamma in this case.
(and would work out similar to double exponential iteration between a base GLM for the mean and a Gamma GLM for the dispersion)

One thing to decide: In the adjusted example, I get WLS estimates for p above 2, and brentq raises because both endpoints have the same sign.
What happens if the estimate for the power parameter is >=2?
In this case the Tweedie model is inconsistent with observing zeros. which means the Tweedie model is not the right model. Two possible solutions: 
- impose the compound Poisson restriction on the power parameter and forcing it to 1.99 or something like that smaller than 2, or
- treat it as a quasi-likelihood model where tweedie is a convenient model to estimated the mean function parameters, and use power > 2, but where the true model is a bit different than Tweedie.

I guess it doesn't make much difference for the mean parameter estimates, but the interpretation will differ.
(E.g. tweedie imposes a restriction on the parameters that relates the Poisson mixing parameters to the gamma claim size parameters, AFAIU)

thequackdaddy commented:
So update for today... 

Promising news and not-so-promising news. 

I realized I had a few mistakes in my prior formula (big surprise). Essentially, I wasn't calculating scale correctly in the psi estimates. I *think* this is correct. 

I'm still unable to match any result from R. I used R's tweedie package to generate random tweedie responses. I then used those reponses to try to use myformula to estimate the variance power. But I couldn't get it to match. I tried a few things... The most clever is as follows... 

R
```R
setwd('c:/workspace')
data <- read.csv('cpunish.csv', sep=",")
library(statmod)
library(tweedie)

data['Response'] <- c(3.16263745520219,2.42027116914422,2.39297856612764,2.06605913761992,1.86598449275048,1.74067160032549,1.2676465565581,1.13566848929499,0.726002078477944,0.384540193573733,0.21490972136221,0.183559419961326,0.131646982996753,0,0,0,0)

tp <- tweedie.profile(Response ~ INCOME + SOUTH - 1, p.vec=seq(1, 2, length=10), data=data, do.plot=TRUE)

tp$p.max

model <- glm(Response ~ INCOME + SOUTH - 1, family=tweedie(var.power=tp$p.max, link.power=0), data=data)
```
Here, `tp$p.max` is 1.380952

Python
```python
import numpy as np
import statsmodels.api as sm

y = np.array([3.16263745520219,2.42027116914422,2.39297856612764,2.06605913761992,1.86598449275048,1.74067160032549,1.2676465565581,1.13566848929499,0.726002078477944,0.384540193573733,0.21490972136221,0.183559419961326,0.131646982996753,0,0,0,0])

model = sm.GLM(y, data.exog[['INCOME', 'SOUTH']],
               family=sm.families.Tweedie(var_power=1.380952, link_power=0))
res = model.fit()
```

Python's `res.summary()` and R's `summary(model)` give very similar results. (yay!). 

But when I try to use my function `model.tweedie_estimate_power(res.mu, 0, 100)`, I get a number around 0.68 when I'd love something around 1.4-1.6.  (boo)

However, and this is why I think this is now becoming promising in my eyes, when I fit the Tweedie model on my actual data, I get a tweedie value of around 1.4, and its very stable. While in the past I had suggested at 1.6 or 1.7 is more appropriate, that's probably for claim types with medical payments and lawyer fees. The particular coverage I fed it is property damage liability only--so the loss dollars are relatively small. (mean of $1,500 or so) and will rarely be more than 5 times the mean. So a 1.4 might be believable. (This is the first time a GLM has been run on this particular book, so I really don't know what to expect.)

thequackdaddy commented:
So good news for today... 

I read through some of the explanations of Taylor's power rule and its similarities to Tweedie. I see how these are very similar now. I also used some of my real data to fit a model. 

I constructed a Tweedie GLM (log link, var_power = 1.38) and got this from from the summary:

```python
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:               LossCost   No. Observations:              1201918
Model:                            GLM   Df Residuals:                   152862
Model Family:                 Tweedie   Df Model:                           23
Link Function:                    log   Scale:                   11423.1560945
Method:                          IRLS   Log-Likelihood:                    nan
Date:                Tue, 19 Apr 2016   Deviance:                   2.3855e+07
Time:                        13:35:05   Pearson chi2:                 1.75e+09
No. Iterations:                    10                                         
```

I then used the Taylor power rule...

```python
model2 = sm.GLM((model.endog - res.mu) ** 2,
                np.column_stack((np.ones(len(res.mu)), np.log(res.mu))),
                family=sm.families.Gamma(sm.families.links.log), freq_weights=model.freq_weights)
```

And I ended up with this...

```python
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                      y   No. Observations:              1201918
Model:                            GLM   Df Residuals:                   152884
Model Family:                   Gamma   Df Model:                            1
Link Function:                    log   Scale:                   42057.8183405
Method:                          IRLS   Log-Likelihood:            -3.1165e+06
Date:                Tue, 19 Apr 2016   Deviance:                   1.8987e+06
Time:                        13:43:12   Pearson chi2:                 6.43e+09
No. Iterations:                    14                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.3367      7.093      1.316      0.188      -4.566      23.239
x1             1.3686      1.465      0.934      0.350      -1.502       4.239
==============================================================================
```

And conveniently... `np.exp(9.3367) = 11347.410738170993` vs the Pearson estimated `11423.1560945`. Additionally, `model.tweedie_estimate_power(res.mu)` results in `1.3616010339176736` vs. `1.3686`. So that makes me think this is working. 

I was able to write a test showing that `model.tweedie_estimate_power` kind of works, but understanding that the dataset is really small, I think the test is biased so I have a huge `rtol` to compensate. (25%).

josef-pkt commented:
That sounds really good.   If GLM-Gamma works also well on real datasets, then we should soon be ready to extend this to double exponential models (for general dispersion modelling).
I will go back to looking at issues and PRs tomorrow.

josef-pkt commented on statsmodels/genmod/families/family.py:
```diff
@@ -1432,3 +1432,193 @@ def resid_anscombe(self, endog, mu):
+        See statsmodels.family.links for more information.
+    var_power : float, optional
+        The variance power.
+    link_power : float, optional
```
shouldn't link_power be part of the link instance, instead of family?

thequackdaddy replied:
Ohh... you stumbled across something I noticed a while back, but never bothered to ask. I was copying from the other classes. I think it should really say `link : a link class, optional`. `link` is actually instantiated in the `__init__` function. 

Reading through statsmodels docs and examples we've always told people to say something like `sm.GLM(endog=y, exog=x, family=sm.families.Gamma(link=sm.families.links.Log))`

Is that the best approach? Personally, I agree with you it seems more natural to pass an instance of a class, not the class itself.

josef-pkt replied:
I didn't compare this yet with the other classes.
We have some other issues open with the limitations of the current implementation.
I'll check this.

thequackdaddy replied:
Sounds good. You might also want to look at [this line](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/families/family.py#L1254) of code. I think we really might want `if link is L.NegativeBinomial:` I don't think that this works as currently written. (Mind you, I don't use Negative Binomial, so I'm not 100% sure here.)

josef-pkt replied:
After an initial browsing through the families and link code:
Your tweedie implementation here follows the same pattern, except that there are no parameterized links yet.
I think this was refactored a long time ago to use only classes instead of instances for links.

I really prefer that we use `link=link.Power(link_power)` in the signature but that needs refactoring of the valid link check

thequackdaddy replied:
Ok. I will proceed with this, then. Perhaps the refactoring can be in a new PR. Hate to break something at this point.

josef-pkt commented on statsmodels/genmod/families/family.py:
```diff
@@ -1432,3 +1432,193 @@ def resid_anscombe(self, endog, mu):
+        -----
+        Not yet...
+        """
+        endog1 = np.clip(endog, 0.1, np.inf)
```
why do we need clipping here, (>=0.1)? Shouldn't this be defined for endog=0 also?

josef-pkt replied:
from the properties of the distribution:
for 1<= p < 2 there will be a mass point at zero.
for p>=2: endog = 0 has zero probability

thequackdaddy replied:
I rethought this... there's some new logic. I still would like to have this to idiot proof cases where they select `p>=2`, but have at least 1 `endog` of 0... I'm using `FLOAT_EPS` instead.

thequackdaddy replied:
Also to explain one thing... 

From "Tweedie Family Densities: Methods of Evaluation" by Dunn and smith I found the following note:

![image](https://cloud.githubusercontent.com/assets/8186432/14838824/e34c4de0-0bea-11e6-9acb-1f640d13048a.png)

I'm using that to determine that when `p=1` and `endog=0`, then `dev =  2 * mu`. (It seems like we've ignored scale historically, so I'm going to stick with that.) The logic becomes...

```python
            dev = np.where(endog == 0,
                           2. * mu,
                           mu - endog * (1 + np.log(mu / endog)))
```
I appreciate thoughts on this trick.

josef-pkt commented on statsmodels/genmod/generalized_linear_model.py:
```diff
@@ -658,6 +658,37 @@ def estimate_scale(self, mu):
 
+    def tweedie_estimate_power(self, mu, low=1.01, high=5.):
+        """
+        Tweedie specific function to estimate scale and shape.
```
mention variance parameter for shape

thequackdaddy replied:
I've reworded this.

josef-pkt commented on statsmodels/genmod/generalized_linear_model.py:
```diff
@@ -658,6 +658,37 @@ def estimate_scale(self, mu):
             raise ValueError("Scale %s with type %s not understood" %
                              (self.scaletype, type(self.scaletype)))
 
+    def tweedie_estimate_power(self, mu, low=1.01, high=5.):
```
add a `method` argument after `mu`, even if it's currently not used.
I think we need to provide, eventually, alternatives to brentq.

thequackdaddy replied:
Done.

josef-pkt commented on statsmodels/genmod/tests/test_glm.py:
```diff
@@ -1318,16 +1365,190 @@ def test_wtd_patsy_missing():
+    # Sample may be too small for this...
+    # assert_allclose(res1.scale, np.exp(res2.params[0]), rtol=0.25)
+    p = model1.tweedie_estimate_power(res1.mu)
+    assert_allclose(p, res2.params[1], rtol=0.25)
```
do you have the estimate for the power in R?   just as reference if it's very different

thequackdaddy replied:
I've tried to run this in R, but it doesn't seem to even run at the moment... for now, I added my R code as a comment... maybe I'm doing something dumb.

thequackdaddy commented:
A few other questions...

* I'd appreciate some double-checking of my `resid_dev` and `deviance` logic. Its a little hairy? 
* Looking at the literature, I think only log and Power are acceptable links for Tweedie. The code *should* work for other links, but I'm not sure what the results would be. (For the records, I only need log.)
* Is there a better name for the variance power estimating function `tweedie_estimate_power`? Maybe `tweedie_pearson_estimate`? Also, is there a way/should we disable/hide this function for non-Tweedie models?

josef-pkt commented on statsmodels/genmod/families/family.py:
```diff
@@ -1432,3 +1432,276 @@ def resid_anscombe(self, endog, mu):
+                           mu - endog * (1 + np.log(mu / endog)))
+        elif p == 2:
+            endog1 = np.clip(endog, FLOAT_EPS, np.inf)
+            dev = -1 + (endog / mu) + np.log(mu / endog1)
```
I think `mu / endog1` is wrong and should be reversed inside the log.
(I was suspicious before about dividing by a possibly zero endog)

checking with Poisson and Gamma families: in both cases it's endog / mu, they also use clipping on the ratio

(Note mu will be > 0 except for degenerate cases, while endog = 0 will be very common in the compound Poisson case.)

I also compared with the SPSS manual, which also has endog / mu.
They have Gamma and Poisson as special cases like you have, and also poisson 2*mu if endog=0

a common problem if endog =0 is that 0 * log(0) needs to be defined as 0, which IIRC is one reason for clipping or adding a tiny value inside the log.

thequackdaddy replied:
Ok. I think actually that formula is "correct" only because `-log(a / b) = log(b/a)`. To clarify this a bit, I reformatted the logic to look like Poisson and Gamma. We should get the same answers (otherwise the tests would fail.)

thequackdaddy replied:
One last thing, I think for endog = 0 we want just mu, not 2 * mu. We multiply by 2 in the last line (in the `return` statement.)

josef-pkt replied:
You're right, I didn't pay attention to the sign.
However, it's usually written endog / mu to avoid ZeroDivisionWarning or Error.

thequackdaddy replied:
Do any of the included datasets have zero and non-zero values? I probably could catch that if I we had that? `cpunish` isn't perfect...

josef-pkt replied:
Another possible change to reduce code duplication.
deviance and deviance residuals are almost the same except for the sum and the sign.
One possibility is to add an internal method that calculates the core computation for each observation, and then let deviance just calculate the weighted sum, and resid_dev calculates the signed sqrt.

josef-pkt commented:
> I'd appreciate some double-checking of my resid_dev and deviance logic. Its a little hairy?

I just added some comments whether the formula is correct, the logic looks fine to me, we need to handle the special cases.

> Looking at the literature, I think only log and Power are acceptable links for Tweedie. The code should work for other links, but I'm not sure what the results would be. (For the records, I only need log.)

see for example #2208
In most cases we (users) want to stick with simple standard link functions, but there are some special cases where we want to use a "weird" (partially inappropriate) link function. There can be problems in those cases, but we don't want to rule them out by hardcoding some restrictions.

Is there a better name for the variance power estimating function tweedie_estimate_power? Maybe tweedie_pearson_estimate? Also, is there a way/should we disable/hide this function for non-Tweedie models?

I think it's fine the way it is, at least at the moment I don't know anything that sounds better.
I'd rather not mention pearson in the name, in case we add other `methods`, e.g. we might want to have extended QMLE (the saddle point approximation) just for comparison or even full MLE if we want an expensive estimate.

I think hiding a method is messy, we don't do it anywhere, and I never tried. We return nan or notimplementederrors in some cases, like loglike here.
I guess in this case we could also estimate the tweedie power parameter based on an initial Poisson or Gamma model, so it's not clear why we would want to raise an exception if the family is different from Tweedie.
Having tweedie in the name and the information in the docstring should be informative enough to users about it's purpose.

josef-pkt commented:
to the last point: It might be that in future we want to have a more general version for the variance parameter estimation, but for now it's just Tweedie. We can always refactor with deprecation if the extention should show up.

one correction:
rename `tweedie_estimate_power` -> `estimate_tweedie_power` in analogy to `estimate_scale` and to have verb in front

josef-pkt commented:
> Do any of the included datasets have zero and non-zero values? I probably could catch that if I we had that? cpunish isn't perfect...

We could use a Poisson count data set. Tweedie wouldn't care if it's integers or not.
But there should also be a dataset used for Tobit (continuous endog with truncation at zero), but I need to find it.

josef-pkt commented:
Tobit uses the `Fair` dataset which has extramarital affairs with many zeros. That's in statsmodels.datasets.
heckman uses female labor supply, but in the PR the data is currently downloaded from the internet.

example usage for fair data is here
https://github.com/statsmodels/statsmodels/pull/2139/files#diff-bfc72f9e2c785755a695d80afb1c090bR54

josef-pkt commented:
BTW: Because Gaussian and Inverse Gaussian are also in the tweedie family, the tweedie_power is (kind of) appropriate for 4 out of 6 families.

thequackdaddy commented:
Ok I'm inching towards the finish line here... I added the fair test, and we're all green. 

As a side note, I tried to also compare the results of `estimate_tweedie_power` with R tweedie's `tweedie.profile`. Our Pearson estimate is getting a 1.763 on the Fair data vs. 1.52449 in R. 

```python
model = sm.GLM(endog=data.endog, exog=data.exog[['rate_marriage', 'age', 'yrs_married']], family=sm.families.Tweedie(var_power=1.5, link_power=0))
res = model.fit()
model.estimate_tweedie_power(res.mu)
```
1.763

```R
model2 <- tweedie.profile(affairs ~ rate_marriage + age + yrs_married - 1, data=data, link.power=0, p.vec=seq(1.1, 1.9, .1),
                         do.plot = TRUE)
model2$p.max
```
1.52449

josef-pkt commented:
sounds good,
about the difference to R `profile`:
Do you know if they use the full profile likelihood or an approximation like the extended quasi-likelihood?
It's not clear which would be more accurate in general. Estimate with profile likelihood can be biased, and assumes correctly specified distribution.
Our mean parameter estimates will be robust to deviations from the tweedie variance function, if we use the pearson estimate this way.

in general: I will look at the latest changes, but also think that we should merge this soon. Further enhancements can be done in future PRs.
Given that you have already the function for the compound poisson tweedie series approximation to the the likelihood, I think we should include the code so it's available for future enhancements. 
I though that a private (leading underscore) module `_tweedie_compound_poisson.py` in genmod would be an appropriate location.
And since I just went looking again for some references, I'd like to get the parameter transformation between tweedie and compound poisson. This would also allow us to simulate a tweedie dataset, and we could do a brief MonteCarlo to check how good the variance parameter estimates are.
Those things could also go in there.

Aside: 
Alternatives to directly using tweedie would include modeling the compound poisson directly, and also extend it to the case when we observe two endog, number of claims and total claim size at the same time. Many of those kind of models would be just combining existing models, instead of adding a new family as in this PR.

thequackdaddy commented:
R's `tweedie.profile` appears to use the infinite series of sums. Its awfully slow. For the `fair` dataset, it took maybe 5 minutes to run. So that's not going to be scalable to the millions. The way it appears to work is that it fits a GLM for each of the selected p's in p.vec then it calculates the loglikelihood. It then plots the fitted log-likelihoods (x-axis is p, y-axis is loglikelihood), fits to a spline, and finds the p which maximizes the spline... so its quite involved. 

I will commit add the gist of the calculation that I provided you as the function you requested. I played with this a little more, but haven't looked at it in a few weeks... and have forgotten much of how it works. It was quite slow (obviously) but there's probably optimizations that are possible with some patience. I won't be adding any tests for the code as I don't really have anything great to check against. 

If you want to add the parameter transformation so we can do some MC sampling, I think it sounds like that's going to be in a future PR for this or do you want included in this PR? 

After you are happy with this, I'm happy to squash.

Thanks so much Josef.

josef-pkt commented:
I think this is a good place to stop and merge, after you added the helper function.

The deviance looks good now, I didn't look at the other things again.
13 commits is not too much, but I'll check if we should do some squashing.

As followup, it would be good to have a notebook that explains a bit the new family, maybe with both test cases (just copied over). AFAIK, we don't have the changelog ready right now, but we will need one or a few paragraph announcing it.

As further followup, we can also check to see if we have all possible results, given that we don't have loglike.  Somewhere we need to check or mention that tweedie can only be estimated by irls and not with the scipy optimizers. (It would be possible to minimize deviance, but the class hierarchy is not set up for maximizing anything else than loglike.)

josef-pkt commented on statsmodels/genmod/tests/test_glm.py:
```diff
@@ -1318,16 +1365,245 @@ def test_wtd_patsy_missing():
+
+
+class TestTweediePower15(CheckTweedie):
+    def __init__(self):
```
convert all the `__init__` to a classmethod `setup_class`   
check other tests for the format

this "should" speed up the unit tests because nose runs `__init__` separately for each method of the class, i.e. this currently estimates the same thing seven times.

josef-pkt commented:
I hadn't realized that you already rebase, looks good, and I don't think it needs any squashing.

However, this adds about a minute or a bit more to the test run on TravisCI.
I guess that using `setup_class` to avoid repeated fitting would speed this up a bit.

thequackdaddy commented:
Yeah I rebased because I think there was conda package update that made it so my build wouldn't work... I noticed you were fighting with it, so I rebased and viola, problem solved. 

I'm now using setupClass when I can. Builds on my machine... let's see what travis thinks.

josef-pkt commented on statsmodels/genmod/tests/test_glm.py:
```diff
@@ -1450,9 +1448,13 @@ def __init__(self):
+        from statsmodels.datasets.cpunish import load_pandas
+        self.data = load_pandas()
+        self.exog = self.data.exog[['INCOME', 'SOUTH']]
+        self.endog = self.data.endog
```
for next time:
one possible trick is to leave these in a method in the super class and call the method directly, which works (IIRC) if the method in the superclass is also a `classmethod`
(could be a `setupClass` or a `getdata` method in super)
super call needs the class twice, IIRC, e.g. `super(TestTweediePower2, TestTweediePower2).get_data()`

("IIRC" means it works but I never remember the details and have to look them up from other examples.)

thequackdaddy replied:
Ahhh... 

I thought about a super but I couldn't get that to work (in about 2 tries)... Didn't attempt this way. 

Thanks for the tip.

josef-pkt commented:
all green, only a little bit faster with the latest versions, and 1 minute or so faster on the older version.

Ready to merge?  From my side it looks ready.

thequackdaddy commented:
I'm happy if you are. 

I'm thinking my next topic is reviewing #2915. Let me re-read some of the papers. 

The full log likelihood will have to wait for another day.

josef-pkt commented:
It's merged. Thanks @thequackdaddy
